rs/crypto/internal/crypto_lib/bls12_381/vetkd/tests/tests.rs

```diff
@@ -1,60 +1,488 @@
 
-        // check that EKS serialization round trips:
-        let eks_bytes = eks.serialize();
-        let eks2 = EncryptedKeyShare::deserialize(eks_bytes).unwrap();
-        assert_eq!(eks, eks2);
+    let node_eks_wrong_did = node_info_wrong_did
+        .iter()
+        .map(|(idx, _pk, eks)| (*idx, eks.clone()))
+        .collect::<Vec<_>>();
 
-        node_info.push((node as u32, node_pk, eks));
+    // With combine_all even if we provide sufficiently many valid shares
+    // if any one share is invalid then combination will fail
+    for rec_threshold in 2..nodes {
+        let mut shares = random_subset(rng, &node_eks, rec_threshold - 1);
+
+        // Avoid using a duplicate index for this test
+        let random_unused_idx = loop {
+            let idx = (rng.gen::<usize>() % node_eks_wrong_did.len()) as u32;
+            if !shares.iter().map(|(i, _eks)| *i).any(|x| x == idx) {
+                break idx as usize;
+            }
+        };
+
+        shares.push(node_eks_wrong_did[random_unused_idx].clone());
+        shares.shuffle(rng);
+
+        let expected_error = if rec_threshold < threshold {
+            EncryptedKeyCombinationError::InsufficientShares
+        } else {
+            EncryptedKeyCombinationError::InvalidShares
+        };
+        assert_eq!(proto.combine_all(&shares), Err(expected_error));
     }
 
-    let ek = EncryptedKey::combine(
-        &node_info,
-        threshold,
-        &master_pk,
-        &tpk,
-        &derivation_path,
-        did,
-    )
-    .unwrap();
+    // Check that duplicate node indexes are detected
+    for rec_threshold in 2..nodes {
+        let mut shares = random_subset(rng, &node_eks, rec_threshold - 1);
+
+        let random_duplicate_idx = loop {
+            let idx = (rng.gen::<usize>() % node_eks.len()) as u32;
 
-    let _k = tsk.decrypt(&ek, &dpk, did).unwrap();
+            if shares.iter().map(|(i, _eks)| *i).any(|x| x == idx) {
+                break idx as usize;
+            }
+        };
+
+        shares.push(node_eks[random_duplicate_idx].clone());
+        shares.shuffle(rng);
+
+        let expected_error = if rec_threshold < threshold {
+            EncryptedKeyCombinationError::InsufficientShares
+        } else {
+            EncryptedKeyCombinationError::DuplicateNodeIndex
+        };
+        assert_eq!(proto.combine_all(&shares), Err(expected_error));
+    }
 
-    let derived_key = tsk
-        .decrypt_and_hash(&ek, &dpk, did, 32, b"aes-256-gcm-siv")
-        .unwrap();
-    assert_eq!(derived_key.len(), 32);
+    // With combine_valid_shares OTOH we detect and reject the invalid shares
+
+    for rec_threshold in threshold..nodes {
+        let mut shares = random_subset(rng, &node_info, rec_threshold);
+        shares.append(&mut random_subset(rng, &node_info_wrong_did, 4));
+        shares.shuffle(rng);
+
+        let combined = proto.combine_valid(&shares);
+        assert!(combined.is_ok(), "Combination succeeded");
```

Comment:
Should the custom panic message here be "Combination failed", rather than "Combination succeeded"?